internal/pkg/manifest/pipeline_stage.go

```diff
@@ -0,0 +1,123 @@
+
+package manifest
+
+import (
+	"fmt"
+
+	"github.com/aws/amazon-ecs-cli-v2/internal/pkg/store/ssm"
+	"github.com/aws/amazon-ecs-cli-v2/internal/pkg/workspace"
+)
+
+// PipelineStage represents configuration for each deployment stage
+// of a workspace. A stage consists of the Archer Environment the pipeline
+// is deloying to and the containerized applications that will be deployed.
+type PipelineStage struct {
+	*associatedEnvironment `yaml:",inline"`
+	LocalApplications      []string `yaml:"-"`
+}
+
+// associatedEnvironment defines the necessary information a pipline stage
+// needs for an Archer Environment.
+type associatedEnvironment struct {
+	// Name of the Archer project this environment belongs to
+	ProjectName string `yaml:"-"`
+
+	// Name of the environment, must be unique within a project.
+	// This is also the name of the pipeline stage.
+	Name string `yaml:"name"`
+	// T region this environment is stored in.
+	Region string `yaml:"-"`
+	// AccountID of the account this environment is stored in.
+	AccountID string `yaml:"-"`
+	// Whether or not this environment is a production environment.
+	Prod bool `yaml:"-"`
+}
+
+// associatedApp defines the name of an Archer application that will be
+// deploye to certain pipeline stage.
+type associatedApp struct {
+	Name string
+}
+
+var fetchAssociatedEnvAndApps func(stageName string) (*associatedEnvironment, []string, error) = func(stageName string) (*associatedEnvironment, []string, error) {
+	// TODO: #239 Centralize the initialization of SSM and workspace into
+	// a common package
+	s, err := ssm.NewStore()
+	if err != nil {
+		return nil, nil, err
+	}
+
+	ws, err := workspace.New()
+	if err != nil {
+		return nil, nil, err
+	}
+
+	summary, err := ws.Summary()
+	if err != nil {
+		return nil, nil, err
+	}
+
+	env, err := s.GetEnvironment(summary.ProjectName, stageName)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	apps, err := ws.AppNames()
+	if err != nil {
+		return nil, nil, err
+	}
+	return &associatedEnvironment{
+		ProjectName: summary.ProjectName,
+		Name:        stageName,
+		Region:      env.Region,
+		AccountID:   env.AccountID,
+		Prod:        env.Prod,
+	}, apps, nil
+}
+
+// NewPipelineStage creates a pipeline stage that maps to the corresponding
+// environment named after the provided `stageName`.
+func NewPipelineStage(stageName string) (*PipelineStage, error) {
+	env, apps, err := fetchAssociatedEnvAndApps(stageName)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create the pipeline stage %s, error: %w",
+			stageName, err)
+	}
+
+	return &PipelineStage{
+		associatedEnvironment: env,
+		LocalApplications:     apps,
+	}, nil
+}
+
+// UnmarshalYAML implements the go-yaml Unmarshaler interface for the type,
+// PipelineStage, such that the associated environment and
+// local applications will be populated after unmarshalling from a pipeline
+// manifest file.
+func (p *PipelineStage) UnmarshalYAML(unmarshal func(interface{}) error) error {
```

Comment:
Should this function be defined as something like `func unmarshal(bytes []byte) (*PipelineStage, error)` instead of a receiver to `PipelineStage`? Seems like it would help remove that recursive problem and remove the need for the `aux` struct?

Having the method attached to a `PipelineStage` means you'd have to have a `PipelineStage` structure that un-marshals into itself? Feels odd.

Reply:
This is outdated since the function is removed. But just to answer your question, we can not change the signature of this method because it's required by the goyaml interface.